zerver/tests/test_message_fetch.py

```diff
@@ -3541,90 +3548,128 @@ def test_invalid_narrow_operand_in_dict(self) -> None:
         # operators
-        invalid_operands = [["1"], [2], None]
         error_msg = 'elem["operand"] is not a string or integer'
+        invalid_operands: List[Dict[str, Any]] = [
+            {"operand": ["1"], "error_msg": error_msg},
+            {"operand": ["2"], "error_msg": error_msg},
+            {"operand": None, "error_msg": error_msg},
+        ]
+
         for operand in ["id", "sender", "channel", "dm-including", "group-pm-with"]:
-            self.exercise_bad_narrow_operand_using_dict_api(operand, invalid_operands, error_msg)
+            self.exercise_bad_narrow_operand_using_dict_api(operand, invalid_operands)
 
         # str or int list is required for "dm" and "pm-with" operator
-        invalid_operands = [None]
+        # First set of invalid operands
         error_msg = 'elem["operand"] is not a string or an integer list'
+        invalid_operands = [{"operand": None, "error_msg": error_msg}]
         for operand in ["dm", "pm-with"]:
-            self.exercise_bad_narrow_operand_using_dict_api(operand, invalid_operands, error_msg)
+            self.exercise_bad_narrow_operand_using_dict_api(operand, invalid_operands)
 
-        invalid_operands = [["2"]]
+        # Second set of invalid operands
         error_msg = 'elem["operand"][0] is not an integer'
+        invalid_operands = [{"operand": ["2"], "error_msg": error_msg}]
         for operand in ["dm", "pm-with"]:
-            self.exercise_bad_narrow_operand_using_dict_api(operand, invalid_operands, error_msg)
+            self.exercise_bad_narrow_operand_using_dict_api(operand, invalid_operands)
 
-        # For others only str is acceptable
-        invalid_operands = [2, None, [1]]
+        # Third set of invalid operands
         error_msg = 'elem["operand"] is not a string'
+        invalid_operands = [
+            {"operand": 2, "error_msg": error_msg},
+            {"operand": None, "error_msg": error_msg},
+            {"operand": [1], "error_msg": error_msg},
+        ]
         for operand in ["is", "near", "has"]:
-            self.exercise_bad_narrow_operand_using_dict_api(operand, invalid_operands, error_msg)
+            self.exercise_bad_narrow_operand_using_dict_api(operand, invalid_operands)
 
         # Disallow empty search terms
         error_msg = 'elem["operand"] cannot be blank.'
-        self.exercise_bad_narrow_operand_using_dict_api("search", [""], error_msg)
+        invalid_operands = [{"operand": "", "error_msg": error_msg}]
+        self.exercise_bad_narrow_operand_using_dict_api("search", invalid_operands)
 
     # The exercise_bad_narrow_operand helper method uses legacy tuple format to
     # test bad narrow, this method uses the current dict API format
     def exercise_bad_narrow_operand_using_dict_api(
-        self, operator: str, operands: Sequence[Any], error_msg: str
+        self, operator: str, operands: Sequence[Dict[str, Any]]
```

Comment:
We should try to avoid the use of `Dict[str, Any]` and switch to things like dataclasses.. This way the type checker always ensures that these helpers are used properly.